# задача — проанализировать базу данных
# В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта

In [284]:
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
 'pwd': 'Sdf4$2;d-d30pp', # пароль
 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
 'port': 6432, # порт подключения
 'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [285]:
pd.options.display.float_format = '{:,.2f}'.format

In [286]:
pd.set_option('display.max_colwidth', 200)

##### округление до двух значений после запятой и вывод до 200 символов в строке.

In [287]:
books=books.reset_index()
ratings=ratings.reset_index()

In [288]:
baza=pd.merge(books, ratings, on='book_id', how='inner')

In [289]:
baza=baza.groupby('author_id').agg({'rating_id': 'count','rating': 'mean'}).reset_index().sort_values('rating_id',ascending=False)

In [290]:
baza.head()

author_id  rating_id  rating
235        236        312    4.29
239        240        166    4.24
553        554        160    3.66
105        106        143    3.74
444        445        111    3.88

# Смотрим базы

In [291]:
query= ''' SELECT *
         FROM books
'''
books = pd.io.sql.read_sql(query, con = engine, index_col = 'book_id')
books.head()

author_id                                                  title  \
book_id                                                                     
1              546                                           'Salem's Lot   
2              465                     1 000 Places to See Before You Die   
3              407    13 Little Blue Envelopes (Little Blue Envelope  #1)   
4               82  1491: New Revelations of the Americas Before Columbus   
5              125                                                   1776   

         num_pages publication_date  publisher_id  
book_id                                            
1              594       2005-11-01            93  
2              992       2003-05-22           336  
3              322       2010-12-21           135  
4              541       2006-10-10           309  
5              386       2006-07-04           268

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

In [292]:
query= ''' SELECT *
         FROM authors
'''
authors = pd.io.sql.read_sql(query, con = engine, index_col = 'author_id')
authors.head()

author
author_id                                
1                              A.S. Byatt
2          Aesop/Laura Harris/Laura Gibbs
3                         Agatha Christie
4                           Alan Brennert
5                Alan Moore/David   Lloyd

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

In [293]:
query= ''' SELECT *
         FROM publishers
'''
publishers = pd.io.sql.read_sql(query, con = engine, index_col = 'publisher_id')
publishers.head()

publisher
publisher_id                                   
1                                           Ace
2                                      Ace Book
3                                     Ace Books
4                                 Ace Hardcover
5             Addison Wesley Publishing Company

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

In [294]:
query= ''' SELECT *
         FROM ratings
'''
ratings = pd.io.sql.read_sql(query, con = engine, index_col = 'rating_id')
ratings.head()

book_id       username  rating
rating_id                                
1                1     ryanfranco       4
2                1  grantpatricia       2
3                1   brandtandrea       5
4                2       lorichen       3
5                2    mariokeller       2

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

In [295]:
query= ''' SELECT *
         FROM reviews
'''
reviews = pd.io.sql.read_sql(query, con = engine, index_col = 'review_id')
reviews

book_id        username  \
review_id                            
1                1    brandtandrea   
2                1      ryanfranco   
3                2        lorichen   
4                3   johnsonamanda   
5                3     scotttamara   
...            ...             ...   
2789           999      martinadam   
2790          1000         wknight   
2791          1000  carolrodriguez   
2792          1000         wendy18   
2793          1000      jarvispaul   

                                                                                                                                                text  
review_id                                                                                                                                             
1                                                         Mention society tell send professor analysis. Over provide race technology continue these.  
2                                                                  Foot glass pretty audience hit themselves. Among admit investment argue security.  
3                   Listen treat keep worry. Miss husband tax but person sport treatment industry. Kitchen decision deep the. Social party body the.  
4                           Finally month interesting blue could nature cultural bit. Prepare beat finish grow that smile teach. Dream me play near.  
5                                   Nation purpose heavy give wait song will. List dinner another whole positive radio fast. Music staff many green.  
...                                                                                                                                              ...  
2789                                                               Later hospital turn easy community. Fact same enter true practice student system.  
2790                                                                                                Change lose answer close pressure. Spend so now.  
2791       Authority go who television entire hair guy population. Pm vote take. Long born close thought person old. Visit ever school manage about.  
2792                                                Or western offer wonder ask. More hear phone foreign amount within cell. Ball poor hot that kid.  
2793                                                 Republican staff bit eat material measure plan. Pretty issue claim rate push energy couple bit.  

[2793 rows x 3 columns]

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

### Посчитайте, сколько книг вышло после 1 января 2000 года

In [296]:
query= ''' SELECT count(book_id) as kolichestvo
         FROM books
         WHERE publication_date>'2000-01-01'
'''

In [297]:
books_21_vek=pd.io.sql.read_sql(query, con = engine,index_col = 'kolichestvo')

In [298]:
books_21_vek

Empty DataFrame
Columns: []
Index: [819]

# Вывод: после 1 января 2000 года вышло 819 книг

### Для каждой книги посчитайте количество обзоров и среднюю оценку

In [299]:
query= '''SELECT BB.book_id, count(DISTINCT BB.review_id) as kolichestvo_obzorov, AVG(BB.rating) as srednaya_ocenka
FROM
(SELECT books.book_id, books.publication_date,  reviews.review_id, ratings.rating
FROM
  books
 
  JOIN reviews
    ON books.book_id = reviews.book_id
 
  JOIN ratings
    ON
      books.book_id = ratings.book_id) AS BB
WHERE BB.publication_date>'2000-01-01'
GROUP BY BB.book_id
ORDER BY BB.book_id ASC
'''

In [300]:
KOL_VO_SRED=pd.io.sql.read_sql(query, con = engine,index_col = 'book_id')

In [301]:
KOL_VO_SRED.head(7)

kolichestvo_obzorov  srednaya_ocenka
book_id                                      
1                          2             3.67
2                          1             2.50
3                          3             4.67
4                          2             4.50
5                          4             4.00
6                          4             3.50
7                          3             3.00

# Вывод: например для пятой книги написано 4 обзора и её средняя оценка - 4

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц

In [302]:
query= '''SELECT MM.publisher_id,MAX(MM.max) as chislo_knig
FROM
(SELECT DD.publisher_id, MAX(DD.count)
FROM
(SELECT AA.publisher_id, COUNT(AA.publisher_id) 
FROM (SELECT   books.num_pages,   publishers.publisher_id
FROM
  books
 
  JOIN publishers
    ON books.publisher_id = publishers.publisher_id) as AA
WHERE AA.num_pages>50
GROUP BY AA.publisher_id) AS DD
GROUP BY DD.publisher_id) AS MM
GROUP BY MM.max, MM.publisher_id
ORDER BY MM.max DESC LIMIT 1
'''

In [303]:
ne_broshurbI=pd.io.sql.read_sql(query, con = engine,index_col = 'publisher_id')

In [304]:
ne_broshurbI

chislo_knig
publisher_id             
212                    42

# Вывод: id издательства, которое которое выпустило наибольшее число книг толще 50 страниц -212(42 книги)

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [305]:
query= ''' select DD.author_id, AVG(DD.avg)
from
(SELECT CC.author_id, AVG(CC.avg)
FROM
(SELECT BB.author_id, BB.book_id,BB.count, BB.avg
FROM
(SELECT AA.author_id, AA.book_id,COUNT(AA.rating_id), AVG(AA.rating)
FROM
(SELECT   books.book_id, books.author_id,  books.num_pages,  ratings.rating, rating_id
FROM
  books
 
  JOIN ratings
    ON books.book_id = ratings.book_id) as AA
GROUP BY AA.author_id, AA.book_id) as BB
WHERE BB.count>=50
GROUP BY BB.author_id, BB.book_id,BB.count, BB.avg) as CC
GROUP BY CC.author_id,CC.avg) as DD
GROUP BY DD.author_id
ORDER BY avg desc
'''

In [306]:
visokaya_avg_ocenka=pd.io.sql.read_sql(query, con = engine)

In [307]:
visokaya_avg_ocenka

author_id  avg
0         236 4.28
1         402 4.26
2         240 4.26
3         376 4.19
4         498 4.08
5         621 3.90
6         235 3.83
7         469 3.79
8         630 3.79
9         106 3.75
10        372 3.75
11        195 3.73
12        554 3.66
13        311 3.62

# Вывод: id автора с самой высокой средней оценкой книг(учитываем только книги с 50 и более оценками) - 236(оценка 4.29)

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [308]:
query= '''SELECT AVG(BB.count)
FROM
(SELECT AA.username, COUNT(AA.review_id)
FROM
(SELECT   ratings.username, COUNT(ratings.rating_id),reviews.review_id

FROM
  ratings
  
  JOIN reviews
    ON ratings.username = reviews.username
GROUP BY ratings.username, reviews.review_id) as AA
WHERE AA.count>50
GROUP BY AA.username) as BB
    '''

In [309]:
visokaya_avg_ocenka=pd.io.sql.read_sql(query, con = engine,index_col = 'avg')

In [310]:
visokaya_avg_ocenka

Empty DataFrame
Columns: []
Index: [24.333333333333332]

# Вывод: среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24